### DISCUSSION TITLE
Relation between runtimes and language server's JDT compiler

### DISCUSSION THREAD

**Discussion by tsaarni** [redhat-developer/vscode-java#2570] [Q&A] [answered]
What are the use cases for runtimes configured in `java.configuration.runtimes` property?

Language server itself runs on [JRE 17 embedded in the extension](https://github.com/redhat-developer/vscode-java/wiki/JDK-Requirements). The project [README](https://github.com/redhat-developer/vscode-java#project-jdks) says

> **Project JDKs**
> If you need to compile your projects against a different JDK version, it's recommended you configure the `java.configuration.runtimes` property in your user settings

but as far as I know, the extension uses JDT Eclipse compier for Java. Not JDK java compiler.

Some guesses:

Is the runtime used when launching the (already compiled) application? 
Is it used to open JDK javadocs and sources when navigating to JDK classes and methods?

**Comment by rgrunber** [comment, marked as answer]
Yes, the specified runtime from `java.configuration.runtimes` can be used when launching the application. It is basically a mapping between the execution environment and a JDK location that can deal with that environment. From there, if (for example), a Maven project is configured for Java 11, then JDT-LS will attempt to locate a JDK that matches that (eg. under the `JavaSE-11` entry).

It does more than just allow you to link to the sources/javadocs. `java.configuration.runtimes` is meant for those wishing to ensure the Java code they write is compatible with a particular version of Java. Imagine using a language feature (eg. Text Blocks, new in Java 15) or API that is from a version of Java that is newer than the version(s) you wish to support. In these cases you will get warnings/errors.

As to the relation with the JDT-LS runtime. We require Java 17 to run JDT-LS, because certain underlying dependencies increased their minimum required version. Not everyone may have Java >= 17 on their system, so it became important to embed a version of Java that could meet this requirement with minimal burden on users. Another advantage of this approach is it can also be used in the event a user has no Java whatsoever. Makes it easier to get started.
